### PR TITLE
fix(collators): clamp auto-detected response template to the generati…

### PR DESCRIPTION
**Pull request by lefft**
# Description

Some chat templates render assistant headers that extend past the generation prompt — harmony-style templates (e.g. `meta-models/Muse-Glimmer-30B`, gpt-oss) render assistant turns as `<|start|>assistant to=user<|message|>…` while the generation prompt is the bare `<|start|>assistant`. When collator templates are auto-detected (`train_target` set, no explicit `response_template`), detection returns the full header, so the trailing ` to=user<|message|>` span is masked out of the loss even though the model must generate it at inference. Heavy fine-tuning corrupts that untrained span: in our Muse-Glimmer FFT runs, the model emitted `' to51'` instead of ` to=user<|message|>51`, leaving 405/500 eval rows unparseable.

**Fix:** after detection, clamp `response_template` to the generation-prompt suffix whenever that suffix is a strict prefix of the detected header — the header remainder then stays in the trained span. The invariant enforced: the masking boundary equals the point where generation begins at inference.

**Blast radius:** none for standard templates. The clamp only fires when the rendered assistant header is strictly longer than the generation prompt; for every template where the two coincide (the normal case) it is a no-op by construction. Verified empirically: all 11 real-tokenizer detection tests (Qwen 2.5/3/3.5/3.6, DeepSeek-R1 ×2, Olmo3, SmolLM 2/3, MiniMax) produce byte-identical results with and without the change. Explicitly-configured `response_template`s are untouched — only auto-detection output changes. New unit test covers the harmony-style case; `meta-models/Muse-Glimmer-30B` added to the integration detection matrix.

**Before/after** (`resolve_collator_templates` on real tokenizers; auto-detection path):

```python
# meta-models/Muse-Glimmer-30B — clamped
before: ('<|start|>assistant to=user<|message|>', '<|eot|>')
after:  ('<|start|>assistant', '<|eot|>')
# trained span before:  '54<|eot|>'                      (header masked — corrupts under FFT)
# trained span after:   ' to=user<|message|>54<|eot|>'   (header supervised)

# google/gemma-4-12B-it — unchanged
before == after: ('<|turn>model', '<turn|>\n')

# Qwen/Qwen3.5-0.8B — unchanged
before == after: ('<|im_start|>assistant', '<|im_end|>\n')
```

Validated end-to-end on the enterprise platform: Muse-Glimmer FFT trained with auto-detected (clamped) templates matches the explicit-template run on banking77 (0.862 vs 0.868 accuracy, 500/500 parseable outputs both).

## Related issues

Supports Muse-Glimmer-30B SFT onboarding (oumi-ai/api#6146).

## Before submitting

- [ ] This PR only changes documentation. (You can ignore the following checks in that case)
- [x] Did you read the [contributor guideline](https://github.com/oumi-ai/oumi/blob/main/CONTRIBUTING.md) Pull Request guidelines?
- [x] Did you link the issue(s) related to this PR in the section above?
- [x] Did you add / update tests where needed?

## Reviewers

At least one review from a member of `oumi-ai/oumi-staff` is required.

### CODE REVIEW

File: src/oumi/builders/collators.py [hunk 1]
```diff
@@ -302,6 +358,9 @@ def resolve_collator_templates(
         header_text=rendered[second_user_end:second_asst],
         eot_ids=eot_ids,
     )
+    response_template = _clamp_to_generation_prompt(
+        tokenizer, response_template, msgs[:-1]
```
Comment (lefft) [inline src/oumi/builders/collators.py]: Here we are referencing `msgs` from the terminated loop above. Works, but somewhat confusing and could be problematic if the loop above is edited in the future and the author doesn't catch this reference. 

Would be cleaner if in the loop above we add e.g. `chosen_msgs = msgs` like:

```py
    for msgs in (msgs_with_sys, msgs_no_sys):
        try:
            rendered = tokenizer.apply_chat_template(
                msgs, tokenize=False, add_generation_prompt=False
            )
            chosen_msgs = msgs
            break
```

and then refer to `chosen_msgs` here.

Comment (lefft) [reply src/oumi/builders/collators.py]: addressed

File: src/oumi/builders/collators.py [hunk 2]
```diff
@@ -302,6 +360,9 @@ def resolve_collator_templates(
         header_text=rendered[second_user_end:second_asst],
         eot_ids=eot_ids,
     )
+    response_template = _clamp_to_generation_prompt(
```
Comment (shanghongsim) [inline src/oumi/builders/collators.py:392]: The clamp approach is clever! I think we could use this as the general approach to detect response prefix `_detect_response_template`.

Comment (shanghongsim) [reply src/oumi/builders/collators.py:392]: It is also unclear when `_clamp_to_generation_prompt` will override `_detect_response_template` with a different result in the current code structure

Comment (lefft) [reply src/oumi/builders/collators.py:392]: Ya fair, to make it clearer, we can explicitly check whether the gen prompt suffix is a proper prefix of `response_template` (as given by `_detect_response_template()`), and then apply the clamp depending on the result of that check. 

Will also add some examples to the docstring similar to what we have for `_detect_eot_template()`.

Comment (shanghongsim) [reply src/oumi/builders/collators.py:392]: sounds like a reasonable approach for now. will approve to unblock.

**Review by shanghongsim — APPROVED** [review]
